torchtitan/utils.py

```diff
@@ -52,6 +52,34 @@ def _warn_overwrite_env(env, val):
     os.environ[env] = val
 
 
+def manual_seed(
+    parallel_dims, world_mesh: DeviceMesh, seed: Optional[int], device: torch.device
+) -> None:
+    """
+    Set the same DTensor manual seed for all ranks within the same DTensor SPMD group, but different
+    seeds across PP groups (if applicable).
+
+    TODO: Link to documentation for DTensor RNG setup? Or explain more here?
+
+    Currently, does not set seeds for the CUDA RNG since TorchTitan always uses DTensor for SPMD parallelisms,
+    and DTensor manages its own RNG tracker, but we could extend to support both if needed.
+    """
+    # to ensure we can control which ranks have same or different seeds, all ranks agree on a starting seed.
+    # if user provides one, we use this. Otherwise rank 0 rolls the dice and everyone else uses that.
+    if seed is None:
+        # Extract the seed for torch's main generator on rank 0 and standardizes on using that to build
+        # seeds for unique SPMD groups
+        seed = torch.get_rng_state()[:8].view(torch.uint64).item()
+        seed_tensor = torch.tensor(seed, device=device, dtype=torch.int64)
+        torch.distributed.broadcast(seed_tensor, src=0)
+        seed = seed_tensor.item()
```

Comment:
Should we worry that users may give different manual seed on different ranks for some reasons (bugs?)?

Reply:
i didn't worry about that, it seems like if you are passing a --seed argument from some launch thing, it would be a lot harder to 'accidentally' get different seeds per rank.  

for the case the user did not pass anything, i think each rank will start with its own seed. so for that case i think we have no choice but to broadcast.